### DISCUSSION TITLE
Feature Request: Flutter Package Support for TipTap Editor

### DISCUSSION THREAD

**Discussion by rameshkumarchandra** [ueberdosis/tiptap#4332] [Feature Requests]
### Description

Description:
We would like to request the addition of Flutter package support for the TipTap editor. Currently, TipTap is a popular rich text editor package for web applications, but there is no official Flutter package available. Adding Flutter support would greatly benefit developers who are building cross-platform applications and looking for a feature-rich rich text editor solution.

Benefits:
- Cross-Platform Compatibility: With Flutter support, developers can easily integrate the TipTap editor into their Flutter applications, enabling them to create consistent rich text editing experiences across both web and mobile platforms.
- Rich Text Editing Capabilities: TipTap provides a wide range of formatting options and tools for creating and editing rich text content. Bringing these capabilities to Flutter would enhance the content creation and editing experience within mobile apps.
- Existing User Base: TipTap already has a strong user base within web development. By adding Flutter support, the package could tap into the Flutter developer community and expand its reach to mobile app development as well.

Proposed Implementation:
We understand that developing a Flutter package involves adapting the existing web-based TipTap editor to work seamlessly within the Flutter framework. This might include creating Flutter widgets that mimic the behavior and features of the web version while taking into consideration the unique aspects of mobile UI/UX.

Additional Considerations:
- Compatibility with Flutter's layout and rendering system.
- Responsive design to accommodate various mobile screen sizes.
- Handling touch interactions and gestures.
- Integration with Flutter's text input and selection mechanisms.

Community Interest:
Given the popularity of Flutter as a cross-platform framework and the demand for feature-rich text editors, we believe that there is a significant interest within the developer community for a Flutter-compatible version of the TipTap editor.

Contributions:
We are willing to contribute to the development of the Flutter package by providing feedback, testing.

We kindly request the development team to consider creating an official TipTap editor package for Flutter, and we look forward to the possibility of having this powerful rich text editor available for cross-platform app development.

Thank you for your time and consideration.

Sincerely,
Ramesh Kumar Chandra Gadi.

### Use Case

In the context of a cross-platform application, the goal is to ensure that the rich content created using the TipTap editor in the web application can be seamlessly replicated with the same behavior and features in the mobile application. This entails enabling users to enjoy a consistent experience across devices, where formatting, styling, media integration, and all other functionalities remain intact when transitioning from the web to the mobile environment. By achieving this consistency, users can effortlessly produce and consume content on both platforms while maintaining a familiar and powerful rich text editing experience.

### Type

New Tiptap API

**Comment by VikingDadMedic** [comment]
I would be willing to help sponsor a Tiptap Flutter integration

**Reply by blackcoffee2** [reply]
I've been working on exactly this i.e. a headless Tiptap engine bridge for Flutter. The PoC is working and shared in the thread. Would love to discuss if you're still interested in sponsoring.

**Comment by FrazeColder** [comment]
Push

**Comment by junoriosity** [comment]
What is the current state of this idea?

**Comment by bdbch** [comment]
We don't have any Flutter developers on our team right now and afaik we don't have any plans for a Flutter version.

Tiptap is built via Prosemirror which uses a lot of web based logic. We have the same issue when it comes to native support for React Native (where a community member built a native bridge).

The biggest issue is that for Tiptap extensions you need to define parseDOM and renderHTML functions which only work on environments where you have said DOM.

How would the parseDOM and renderHTML configurations for extensions work in Flutter / React Native environments? Would extension developers **always** be required to write multiple variants of said functions for other environments?

**Comment by bdbch** [comment]
@ThisDavidRichard I'm right with no Flutter plans right? :)

I think for now the best solution for native / mobile is a webview sadly.

**Reply by FullstackWEB-developer** [reply]
Sure, It's not a perfect solution, but I think it's a negotiable proposal.

**Comment by yoga-teachedison** [comment]
How about converting the data into an HTML and render it in Flutter App using something like **flutter_widget_from_html**

**Reply by FullstackWEB-developer** [reply]
I tried it but it wasn't as good as webview

**Comment by eodAmalitech** [comment]
I am a flutter developer and I am willing to support and push this, @VikingDadMedic how far with this idea.

**Reply by bdbch** [reply]
Would really appreciate if something is happening in this regard if you could link your flutter library here so other people can find and use it!

**Comment by aldrinjenson** [comment]
Hi, is there any update here? 
Feeling the same requirements!

**Reply by arnaugomez** [reply]
Hi Aldrin. Here is our roadmap:

https://tiptap.dev/roadmap

What would be your specific requirements for Flutter/mobile support? If you have specific requirements, let us know. That will help us shape the roadmap.

As of now, native mobile support (using native framework elements instead of a WebView) would take significant time and resources to develop. Besides, there are many target platforms: react native, Flutter, native iOS, native Android, etc. So before we start working on this, we would need to know what platforms are the most requested. That's why we're currently focused on other projects. @ThisDavidRichard is our product guy who takes care of the roadmap, maybe he can share more info on this.

**Reply by bdbch** [reply]
To bring in my 2 cents - I think we as maintainers should not spend too much time spreading our work across multiple frameworks, libraries and languages. While it would be cool to have Flutter (or native RN) support, I wonder if the amount of work required to pull this off is worth the numbers of people actually implementing it in Flutter/RN.

The biggest problem still is that Tiptap is very dependent on the DOM, the EditorView uses the browser DOM, the parser and serializer is using the Browser DOM, there are a lot of coupled things that would take a massive undertaking to get this to work outside of the browser.

**Comment by blackcoffee2** [comment]
I've been working on a solution to this. Sharing it here since it directly addresses the DOM dependency problem raised in this thread.

**The approach:** Instead of porting ProseMirror to Dart, run the original Tiptap JavaScript inside an invisible OS WebView (WKWebView on iOS, Chromium WebView on Android). The WebView is a computation engine only i.e. zero pixels from it reach the screen. All rendering, input handling, and UI are native Flutter.

**How it works:** The engine creates a real Tiptap v3 editor mounted on a hidden DOM element. On every transaction, it serializes the document into annotated JSON with ProseMirror position offsets on every node and pushes it to Flutter through a JSON message protocol. Flutter sends commands back (`insertText`, `toggleBold`, `setContent`, etc.) the same way.

**What exists today:**

- **tiptap-engine** A headless Tiptap runtime with a message-based protocol. Includes schema introspection so ports can auto-generate toolbars from the loaded extensions. Tested and building.
- A **Flutter proof-of-concept** app that loads the engine, sends commands, and receives state updates. The bridge works end-to-end on Android i.e. tested. Should work on iOS without any issues. Engine loads in under a second.

The engine is platform-agnostic — the same bundle and protocol could be used to build Swift, Kotlin, or React Native ports.

Repo: [[LINK]](https://github.com/blackcoffee2/tiptap-engine)

Proof of concept flutter repo link: [[LINK]](https://github.com/blackcoffee2/tiptap-engine-flutter-app)

Feedback welcome, especially from anyone who has tried the visible-WebView approach and run into the keyboard/scroll/focus issues that motivated this architecture.

**Comment by blackcoffee2** [comment]
An alpha release of my flutter port is now on pub.dev: [tiptap_flutter](https://pub.dev/packages/tiptap_flutter)

**Reply by AnithaB05** [reply]
Hi, Great work on bringing Tiptap to Flutter. 

I wanted to ask if you have any plans or ideas to add Flutter Web support for tiptap_flutter in the future.

Since Tiptap itself is browser/DOM-oriented, I was curious whether a Flutter Web implementation is technically feasible with the current architecture, or if there are major blockers around ProseMirror/DOM integration.

**Reply by blackcoffee2** [reply]
Flutter web support for Tiptap will be interesting.
I'll have a look.
Can you open an issue at [tiptap_flutter](https://github.com/blackcoffee2/tiptap-flutter) repo?

**Reply by AnithaB05** [reply]
Thanks! I’ve opened an issue in the tiptap_flutter repo.

**Comment by blackcoffee2** [comment]
Update: tiptap_flutter is now at v0.0.4 with image insertion support, improved rendering, and a cleaner API. No longer alpha.

pub.dev: https://pub.dev/packages/tiptap_flutter
GitHub: https://github.com/blackcoffee2/tiptap-flutter
